quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/tracing/OTelTracingProvider.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.quarkus.runtime.tracing;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.semconv.ExceptionAttributes;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.jboss.logging.Logger;
+import org.keycloak.tracing.TracingProvider;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Tracing provider leverages OpenTelemetry Tracing
+ */
+public class OTelTracingProvider implements TracingProvider {
+    private static final Logger log = Logger.getLogger(OTelTracingProvider.class);
+    private final OpenTelemetry openTelemetry;
+    private final Deque<Scope> scopes;
+
+    public OTelTracingProvider(OpenTelemetry openTelemetry) {
+        this.openTelemetry = openTelemetry;
+        this.scopes = new ArrayDeque<>();
+    }
+
+    OpenTelemetry getOpenTelemetry() {
+        return openTelemetry;
+    }
+
+    @Override
+    public Span getCurrentSpan() {
+        return Span.current();
+    }
+
+    @Override
+    public Span startSpan(String tracerName, String spanName) {
+        var tracer = getTracer(tracerName);
+        return startSpan(tracer.spanBuilder(spanName));
+    }
+
+    @Override
+    public Span startSpan(SpanBuilder builder) {
+        var currentSpan = builder.startSpan();
+        scopes.push(currentSpan.makeCurrent());
+
+        if (log.isDebugEnabled()) {
+            log.debugf("Start span '%s' (spanId: '%s')", getSpanName(currentSpan), currentSpan.getSpanContext().getSpanId());
+        }
+
+        return currentSpan;
+    }
+
+    @Override
+    public void endSpan() {
+        if (scopes.isEmpty()) {
+            log.info("All spans during tracing already ended. Verify that 'TracingProvider.endSpan()' method was not executed multiple times in the span hierarchy.");
+            return;
+        }
+
+        var span = getCurrentSpan();
```

Comment:
Each session has it's own `OTelTracingProvider`, right?
Just to make sure the scopes in the `Deque` don't belong to different sessions.

Reply:
@brunobat This is a very good point. In the initial phase, I made the provider application scoped, but it should not be the case when the state (scopes) is present - it should be request/session scoped. I'll send an additional PR. 

Is an `OpenTelemetry` bean considered application-scoped and is not changed in runtime? The thing is, whether I should request the bean from Arc every single request, or use the same instance all the time?

Reply:
`OpenTelemetry` is safe, it's application scoped and will not change.
The problem is with the scopes you are enqueuing and dequeuing and make sure they are from the same request (I called it session before). They need to be managed in a thread safe way.

Reply:
Addressed in: https://github.com/keycloak/keycloak/pull/35428